### DISCUSSION TITLE
Mock S3 "Custom" endpoint/region for local testing purposes?

### DISCUSSION THREAD

**Discussion by brainstorm** [awslabs/aws-sdk-rust#253] [Q&A] [answered]
Hi aws-sdk-rust team,

From what I see on [the defined S3 endpoints](https://github.com/awslabs/aws-sdk-rust/blob/main/sdk/s3/src/aws_endpoint.rs), there does not seem to be any way to define a non-production S3 endpoint for local (mock) testing?

I'm using the excellent [s3-server](https://github.com/datenlord/s3-server) crate [from @datenlord on my project's S3-related testing operations](https://github.com/umccr/htsget-rs/blob/684270b4ebdeaedf71d0f1116a91c4e1bfa63674/htsget-search/src/storage/aws.rs#L244-L285) but I can't seem to be able to point aws-sdk-rust's S3 client to `http://localhost:8041` as I need to mock test some of my functionality?

Rusoto does have such a facility via "custom Regions", see: https://github.com/rusoto/rusoto/pull/759.

EDIT: while I was writing this and checking out this [S3 signing test](https://github.com/awslabs/aws-sdk-rust/blob/main/sdk/s3/tests/signing-it.rs), I noticed `smithy-client`, which looks like a lower-level way to test things out and perhaps not requiring `s3-server` crate at all... I'm not sure it supports head_object or similar but I'll check that, but please let me know if there's a better way to test those ;)

**Comment by rcoh** [comment, marked as answer]
Hello! using s3-server is a totally reasonable testing strategy, you can accomplish by setting a custom endpoint:
(note that you'll need to a dependency on `http` directly and construct your client a little differently):
```rust
use aws_sdk_s3::Client;
use aws_sdk_s3::Endpoint;
use http::Uri;

#[tokio::main]
async fn main() {
    let conf = aws_config::load_from_env().await;
    let ep = Endpoint::immutable(Uri::from_static("http://localhost:8543"));
    let s3_conf = aws_sdk_s3::config::Builder::from(&conf)
        .endpoint_resolver(ep)
        .build();
    let s3 = Client::from_conf(s3_conf);
    let buckets = s3.list_buckets().send().await;
    println!("got buckets: {:#?}", buckets);
}
```

You can also use the hyper service as a connector directly (which I think is pretty cool, personally!):

```rust
use aws_sdk_s3::ByteStream;
use s3_server::storages::fs::FileSystem;
use s3_server::S3Service;
use smithy_http::result::ConnectorError;
use std::error::Error;
use std::future::Future;
use std::pin::Pin;

use aws_sdk_s3::Client;
use aws_sdk_s3::Endpoint;
use http::Uri;
use s3_server::SharedS3Service;
use smithy_http::body::SdkBody;

#[derive(Clone)]
struct AwsS3Server(SharedS3Service);

impl tower::Service<http::Request<SdkBody>> for AwsS3Server {
    type Response = http::Response<SdkBody>;
    type Error = ConnectorError;

    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;

    fn poll_ready(
        &mut self,
        cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<Result<(), Self::Error>> {
        self.0.poll_ready(cx).map_err(ConnectorError::io)
    }

    fn call(&mut self, req: http::Request<SdkBody>) -> Self::Future {
        let req = req.map(|sdkbody| hyper::Body::wrap_stream(ByteStream::new(sdkbody)));
        let inner = self.0.call(req);
        let fut = async move {
            let resp = inner.await;
            resp.map(|ok| ok.map(SdkBody::from))
                .map_err(ConnectorError::io)
        };
        Box::pin(fut)
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let conf = aws_config::load_from_env().await;
    let ep = Endpoint::immutable(Uri::from_static("http://localhost:8543"));
    let s3_conf = aws_sdk_s3::config::Builder::from(&conf).build();
    let fs = FileSystem::new("./s3")?;
    // setup the service
    let server = AwsS3Server(S3Service::new(fs).into_shared());
    let s3 = Client::from_conf_conn(s3_conf, server);
    let buckets = s3.list_buckets().send().await;
    println!("got buckets: {:#?}", buckets);
    Ok(())
}
```

**Reply by brainstorm** [reply]
Excellent! Thanks so much for the detailed explanation! ;)

**Reply by brainstorm** [reply]
Hello again @rcoh, fast forward to 2023, we'd like to remove the `s3-server` dependency since it has OpenSSL and Rusoto dependencies we'd really like to remove from our codebase... and ideally not requiring a localhost server running to respond to requests.

Unfortunately it's not entirely clear to me [how to mock a few methods](https://github.com/umccr/htsget-rs/issues/131#issuecomment-1375168773) with [the official AWS trait mock docs](https://docs.aws.amazon.com/sdk-for-rust/latest/dg/testing.html).

This is a followup so folks can land on [this issue and improve docs and support for Client mocking](https://github.com/awslabs/aws-sdk-rust/issues/199#issuecomment-1386430875), so replies can be followed up there instead?
